torch/csrc/autograd/saved_variable.cpp

```diff
@@ -96,8 +88,8 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
   // if versions don't match
   auto grad_fn = saved_original ? data_.grad_fn()
                                 : is_inplace_view_ ? weak_grad_fn_.lock()
-                                                   : grad_fn_;
-  if (has_grad_fn_ && !grad_fn) {
+                                                   : nullptr;
+  if (!saved_original && !grad_fn) {
```

Comment:
The logic is:
If you saved the original variable, you never want to look at what's in `saved_for`.
Now if you did not save the original variable, it can not be a leaf, so `has_grad_fn` is always true.